xls/passes/rematerialization_pass.cc

```diff
@@ -388,6 +434,87 @@ FindRematerializationOpportunities(
   return result;
 }
 
+// TODO(zihao): replace `FindRematerializationOpportunitiesAtNode` later
+absl::StatusOr<RematEdgeOpportunity>
+FindRematerializationOpportunitiesAtEdge(
+    Node* src, Node* dst, FunctionBase* f,
+    const absl::flat_hash_map<Node*, int64_t>& schedule,
+    const InverseSchedule& inverse_schedule) {
+  int64_t dst_stage_num = schedule.at(dst);
+  absl::flat_hash_set<Node*> available;
+  for (Node* node : inverse_schedule[dst_stage_num]) {
+    available.insert(node);
+  }
+
+  // TODO: remove nodes that would make the schedule worse from `available`
+
+  absl::flat_hash_set<Node*> unavailable;
+  for(int64_t i = 0; i < dst_stage_num; ++i) {
+    for(const auto& node : inverse_schedule[i]){
+      unavailable.insert(node);
+    }
+  }
+
+  absl::flat_hash_set<Node*> chunk = ComputeDeadNodeChunk(unavailable, src);
+  XLS_CHECK(!chunk.empty());
+
+  absl::flat_hash_map<Node*, int64_t> topo_index;
+
+  {
+    std::vector<Node*> topo_sort = TopoSort(f).AsVector();
+    for (int64_t i = 0; i < topo_sort.size(); ++i) {
+      topo_index[topo_sort[i]] = i;
+    }
+  }
+
+  absl::flat_hash_map<Node*, Node*> clones;
+  std::vector<Node*> chunk_toposorted(chunk.begin(), chunk.end());
+  std::stable_sort(chunk_toposorted.begin(), chunk_toposorted.end(),
+                  [&](Node* x, Node* y) -> bool {
+                    return topo_index.at(x) < topo_index.at(y);
+                  });
+  for (Node* node : chunk_toposorted) {
+    std::vector<Node*> cloned_operands;
+    for (Node* operand : node->operands()) {
+      cloned_operands.push_back(clones.at(operand));
+    }
+    XLS_ASSIGN_OR_RETURN(clones[node], node->Clone(cloned_operands));
+  }
+  
+  double quality = (schedule.at(dst) - schedule.at(src)) *
```

Comment:
It seems like what you're doing is copying the entire chunk to the stage of `dst`, but the chunk may have inputs like Params that cannot be copied (consider what happens if you have a 2 stage pipeline and a param on stage 2, if the stream of inputs is {1, 2, 3} then when stage 1 is computing on value 3 then stage 2 needs to be fed value 2 which means that the Param needs to be flopped / fed in from stage 1).

To fix this you need to subtract off any flops incurred by these inputs. More generally, you don't need to take the entire `chunk`, you could cut off at some earlier point and treat the leaf nodes as inputs. Maybe a min cut (as in the old scheduler) could be used for this?

Reply:
I think your initial implementation also copies the entire chunk if I haven't misunderstood it. I wonder if the intention of `if (std::any_of(chunk.begin(), chunk.end(),...` line is to make sure the chunk doesn't contain params.

To avoid copying the entire chunk, a min-cut may be a suitable way to model it. It intuitively makes sense. Extra efforts will be needed to figure out the correct cost function.

I guess this method is talking about the same as the 3rd recomputing method in XLS's [Remat issue](https://github.com/google/xls/issues/544).

Reply:
I think you're right about the intention behind the `if`, but I think it's still wrong and the core issue is actually that our notion of availability (as I just commented above) is impoverished. With the current definition, anything in previous stages is going to be unavailable, so of course ComputeDeadNodeChunk will terminate at params and literals. But if we adopt a notion of availability based on use-def ranges, then it's possible to find available params.

The min cut thing seems interesting, but maybe a distraction for now. And yes, it is related to #3.